NetKet/Machine/activations.hpp

```diff
@@ -0,0 +1,192 @@
+};
+
+inline double lncosh(double x) {
+  const double xp = std::abs(x);
+  if (xp <= 12.) {
+    return std::log(std::cosh(xp));
+  } else {
+    const static double log2v = std::log(2.);
+    return xp - log2v;
+  }
+}
+
+// ln(cos(x)) for std::complex argument
+// the modulus is computed by means of the previously defined function
+// for real argument
+inline std::complex<double> lncosh(std::complex<double> x) {
+  const double xr = x.real();
+  const double xi = x.imag();
+
+  std::complex<double> res = lncosh(xr);
+  res += std::log(
+      std::complex<double>(std::cos(xi), std::tanh(xr) * std::sin(xi)));
+
+  return res;
+}
+
+class Identity : public AbstractActivation {
+ private:
+  using VectorType = Eigen::Matrix<std::complex<double>, Eigen::Dynamic, 1>;
+
+ public:
+  // A = Z
+  inline void operator()(const VectorType &Z, VectorType &A) {
+    A.noalias() = Z;
+  }
+
+  // Apply the (derivative of activation function) matrix J to a vector F
+  // A = Z
+  // J = dA / dZ = I
+  // G = J * F = F
+  inline void ApplyJacobian(const VectorType & /*Z*/, const VectorType & /*A*/,
+                            const VectorType &F, VectorType &G) {
+    G.noalias() = F;
+  }
+};
+
+class Lncosh : public AbstractActivation {
+ private:
+  using VectorType = Eigen::Matrix<std::complex<double>, Eigen::Dynamic, 1>;
+
+ public:
+  // A = Lncosh(Z)
+  inline void operator()(const VectorType &Z, VectorType &A) {
+    for (int i = 0; i < A.size(); ++i) {
+      A(i) = lncosh(Z(i));
+    }
+  }
+
+  // Apply the (derivative of activation function) matrix J to a vector F
+  // A = Lncosh(Z)
+  // J = dA / dZ
+  // G = J * F
+  inline void ApplyJacobian(const VectorType &Z, const VectorType & /*A*/,
+                            const VectorType &F, VectorType &G) {
+    G.array() = F.array() * Z.array().tanh();
+  }
+};
+
+class Tanh : public AbstractActivation {
+ private:
+  using VectorType = Eigen::Matrix<std::complex<double>, Eigen::Dynamic, 1>;
+
+ public:
+  // A = Tanh(Z)
+  inline void operator()(const VectorType &Z, VectorType &A) {
+    A.array() = Z.array().tanh();
+  }
+
+  // Apply the (derivative of activation function) matrix J to a vector F
+  // A = Tanh(Z)
+  // J = dA / dZ
+  // G = J * F
+  inline void ApplyJacobian(const VectorType & /*Z*/, const VectorType &A,
+                            const VectorType &F, VectorType &G) {
+    G.array() = F.array() * (1 - A.array() * A.array());
+  }
+};
+
+class Activation : public AbstractActivation {
+  using Ptype = std::unique_ptr<AbstractActivation>;
+
+  Ptype m_;
+
+  int mynode_;
+
+ public:
+  using VectorType = typename AbstractActivation::VectorType;
+
+  explicit Activation(const json &pars) { Init(pars); }
+  void Init(const json &pars) {
+    CheckInput(pars);
+    MPI_Comm_rank(MPI_COMM_WORLD, &mynode_);
+    if (pars["Activation"] == "Lncosh") {
+      m_ = Ptype(new Lncosh());
+      if (mynode_ == 0) {
+        std::cout << "# # Activation: "
```

Comment:
Use `InfoMessage` (or maybe `DebugMessage`, as this output is probably only useful for debugging) instead of `std::cout` here.